providers/class.two-factor-backup-codes.php

```diff
@@ -0,0 +1,274 @@
+<?php
+/**
+ * Class for creating a backup codes provider.
+ *
+ * @since 0.1-dev
+ *
+ * @package Two_Factor
+ */
+class Two_Factor_Backup_Codes extends Two_Factor_Provider {
+
+	/**
+	 * The user meta backup codes key.
+	 * @type string
+	 */
+	const BACKUP_CODES_META_KEY = '_two_factor_backup_codes';
+
+	/**
+	 * The number backup codes.
+	 * @type int
+	 */
+	const NUMBER_OF_CODES = 10;
+
+	/**
+	 * Ensures only one instance of this class exists in memory at any one time.
+	 *
+	 * @since 0.1-dev
+	 */
+	static function get_instance() {
+		static $instance;
+		$class = __CLASS__;
+		if ( ! is_a( $instance, $class ) ) {
+			$instance = new $class;
+		}
+		return $instance;
+	}
+
+	/**
+	 * Class constructor.
+	 *
+	 * @since 0.1-dev
+	 */
+	protected function __construct() {
+		add_action( 'two-factor-user-options-' . __CLASS__, array( $this, 'user_options' ) );
+		add_action( 'admin_notices', array( $this, 'admin_notices' ) );
+		add_action( 'wp_ajax_two_factor_backup_codes_generate', array( $this, 'ajax_generate_json' ) );
+
+		return parent::__construct();
+	}
+
+	/**
+	 * Displays an admin notice when backup codes have ran out.
+	 *
+	 * @since 0.1-dev
+	 */
+	public static function admin_notices() {
+		// Only show this notice if we are out of backup codes.
+		$user_id = get_current_user_id();
+		$backup_codes = get_user_meta( $user_id, self::BACKUP_CODES_META_KEY, true );
+		if ( ! empty( $backup_codes ) ) {
+			return;
+		}
+
+		// Only show when the provider is enabled.
+		if ( false === Two_Factor_Backup_Codes::get_instance()->is_enabled( $user_id ) ) {
```

Comment:
Why are we instantiating this class just to call what could just as easily be a static method?

I'd sooner see `is_enabled()` made static, so we can just call `self::is_enabled()`